src/runtime_src/core/pcie/linux/shim.cpp

```diff
@@ -1464,7 +1464,25 @@ int shim::xclLoadAxlf(const axlf *buffer)
     axlf_obj.kds_cfg.slot_size = mCoreDevice->get_ert_slots(xml_data, xml_size).second;
 
     int ret = mDev->ioctl(mUserHandle, DRM_IOCTL_XOCL_READ_AXLF, &axlf_obj);
-    if(ret)
+    if (ret && errno == EAGAIN) {
+        //special case for aws
+        //if EAGAIN is seen, that means a pcie removal&rescan is ongoing, let's just
+        //wait and reload 2nd time -- this time the there will be no device id
+        //change, hence no pcie removal&rescan, anymore
+        int dev_hotplug_done = 0;
+        std::string err;
+        dev_fini();
```

Comment:
Without really being able to say why, I don't think dev_ini() and  dev_fini() are safe outside constructor and destructor.  I have never seen these called arbitrarily as is done here.

Reply:
there is a hotplug happening here, so we have to close the opened device otherwise the unloading of the driver will hang.
same logic in this file also applies to reset, p2p config
I will add more comments here